File: web/src/lib/api/demands.ts
/**
 * API client for club demands and demand matching
 */

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || "http://127.0.0.1:8000";

// TypeScript interfaces

export interface MatchCounts {
  suggested: number;
  proposed: number;
  interested: number;
  rejected: number;
  signed: number;
}

export interface DemandSummary {
  id: number;
  club_name: string;
  position: string[];
  status: "open" | "closed";
  priority: "low" | "medium" | "high";
  deadline: string | null;
  created_date: string;
  match_counts: MatchCounts;
}

export interface DemandDetail {
  id: number;
  club_name: string;
  club_contact_name: string | null;
  club_contact_email: string | null;
  created_date: string;
  deadline: string | null;
  status: "open" | "closed";
  priority: "low" | "medium" | "high";
  position: string[];
  age_min: number | null;
  age_max: number | null;
  nationality_preferences: string[] | null;
  budget_min: number | null;
  budget_max: number | null;
  metrics_requirements: Record<string, { min?: number; max?: number }> | null;
  playing_style_notes: string | null;
  default_season: string;
  internal_notes: string | null;
  match_counts: MatchCounts;
}

export interface PlayerInfo {
  id: number;
  name: string;
  team: string | null;
  league: string | null;
  season: string | null;
  position: string | null;
  age: number | null;
  nationality: string | null;
  image_url: string | null;
  market_value: number | null;
  key_metrics: Record<string, number>;
}

export interface DemandMatch {
  id: number;
  demand_id: number;
  player_season_id: number;
  match_score: number;
  status: "suggested" | "proposed" | "interested" | "rejected" | "signed";
  added_date: string;
  status_updated_date: string;
  notes: string | null;
  player: PlayerInfo;
}

export interface CreateDemandRequest {
  club_name: string;
  club_contact_name?: string;
  club_contact_email?: string;
  deadline?: string;
  priority?: "low" | "medium" | "high";
  position: string[];
  age_min?: number;
  age_max?: number;
  nationality_preferences?: string[];
  budget_min?: number;
  budget_max?: number;
  metrics_requirements?: Record<string, { min?: number; max?: number }>;
  playing_style_notes?: string;
  default_season: string;
  internal_notes?: string;
}

export interface UpdateDemandRequest {
  club_name?: string;
  club_contact_name?: string;
  club_contact_email?: string;
  deadline?: string;
  status?: "open" | "closed";
  priority?: "low" | "medium" | "high";
  position?: string[];
  age_min?: number;
  age_max?: number;
  nationality_preferences?: string[];
  budget_min?: number;
  budget_max?: number;
  metrics_requirements?: Record<string, { min?: number; max?: number }>;
  playing_style_notes?: string;
  default_season?: string;
  internal_notes?: string;
}

export interface UpdateMatchRequest {
  status?: "suggested" | "proposed" | "interested" | "rejected" | "signed";
  notes?: string;
}

export interface RunMatchingRequest {
  season?: string;
  limit?: number;
}

// API functions

/**
 * Create a new club demand
 */
export async function createDemand(
  data: CreateDemandRequest
): Promise<DemandDetail> {
  const response = await fetch(`${API_BASE_URL}/demands`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(data),
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error.detail || "Failed to create demand");
  }

  return response.json();
}

/**
 * List all demands with optional filters
 */
export async function listDemands(params?: {
  status?: "open" | "closed";
  club_name?: string;
  position?: string;
  date_from?: string;
  date_to?: string;
}): Promise<DemandSummary[]> {
  const query = new URLSearchParams();
  if (params?.status) query.append("status", params.status);
  if (params?.club_name) query.append("club_name", params.club_name);
  if (params?.position) query.append("position", params.position);
  if (params?.date_from) query.append("date_from", params.date_from);
  if (params?.date_to) query.append("date_to", params.date_to);

  const url = `${API_BASE_URL}/demands${query.toString() ? `?${query}` : ""}`;
  const response = await fetch(url);

  if (!response.ok) {
    throw new Error("Failed to fetch demands");
  }

  return response.json();
}

/**
 * Get demand detail
 */
export async function getDemand(demandId: number): Promise<DemandDetail> {
  const response = await fetch(`${API_BASE_URL}/demands/${demandId}`);

  if (!response.ok) {
    if (response.status === 404) {
      throw new Error("Demand not found");
    }
    throw new Error("Failed to fetch demand");
  }

  return response.json();
}

/**
 * Update a demand
 */
export async function updateDemand(
  demandId: number,
  data: UpdateDemandRequest
): Promise<DemandDetail> {
  const response = await fetch(`${API_BASE_URL}/demands/${demandId}`, {
    method: "PATCH",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(data),
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error.detail || "Failed to update demand");
  }

  return response.json();
}

/**
 * Delete a demand
 */
export async function deleteDemand(demandId: number): Promise<void> {
  const response = await fetch(`${API_BASE_URL}/demands/${demandId}`, {
    method: "DELETE",
  });

  if (!response.ok) {
    throw new Error("Failed to delete demand");
  }
}

/**
 * Run matching algorithm for a demand
 */
export async function runMatching(
  demandId: number,
  params?: RunMatchingRequest
): Promise<DemandMatch[]> {
  const response = await fetch(`${API_BASE_URL}/demands/${demandId}/match`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(params || {}),
  });

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error.detail || "Failed to run matching");
  }

  return response.json();
}

/**
 * Get all matched players for a demand
 */
export async function getDemandMatches(
  demandId: number,
  params?: {
    status?: "suggested" | "proposed" | "interested" | "rejected" | "signed";
    sort_by?: "match_score" | "added_date";
  }
): Promise<DemandMatch[]> {
  const query = new URLSearchParams();
  if (params?.status) query.append("status", params.status);
  if (params?.sort_by) query.append("sort_by", params.sort_by);

  const url = `${API_BASE_URL}/demands/${demandId}/matches${
    query.toString() ? `?${query}` : ""
  }`;
  const response = await fetch(url);

  if (!response.ok) {
    throw new Error("Failed to fetch demand matches");
  }

  return response.json();
}

/**
 * Update match status and/or notes
 */
export async function updateMatch(
  demandId: number,
  playerSeasonId: number,
  data: UpdateMatchRequest
): Promise<DemandMatch> {
  const response = await fetch(
    `${API_BASE_URL}/demands/${demandId}/matches/${playerSeasonId}`,
    {
      method: "PATCH",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(data),
    }
  );

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error.detail || "Failed to update match");
  }

  return response.json();
}

/**
 * Propose a player to the club (shortcut to set status to proposed)
 */
export async function proposeMatch(
  demandId: number,
  playerSeasonId: number
): Promise<DemandMatch> {
  const response = await fetch(
    `${API_BASE_URL}/demands/${demandId}/matches/${playerSeasonId}/propose`,
    {
      method: "POST",
    }
  );

  if (!response.ok) {
    const error = await response.json().catch(() => ({}));
    throw new Error(error.detail || "Failed to propose match");
  }

  return response.json();
}

/**
 * Remove a player from demand matches
 */
export async function deleteMatch(
  demandId: number,
  playerSeasonId: number
): Promise<void> {
  const response = await fetch(
    `${API_BASE_URL}/demands/${demandId}/matches/${playerSeasonId}`,
    {
      method: "DELETE",
    }
  );

  if (!response.ok) {
    throw new Error("Failed to delete match");
  }
}

File: scouting/api/routers/demands.py
"""API router for club demands and demand matching."""

from __future__ import annotations

from datetime import datetime
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy import and_, func
from sqlalchemy.orm import Session, joinedload

from scouting.api.services.demand_matching import (
    create_demand_matches,
    find_matching_players,
)
from scouting.db.models import (
    ClubDemand,
    DemandMatch,
    League,
    PlayerSeason,
    Season,
    Team,
)


def get_db():
    """Yield a database session for request lifetime."""
    from scouting.db.session import get_session_factory

    factory = get_session_factory()
    db = factory()
    try:
        yield db
    finally:
        db.close()


router = APIRouter(prefix="/demands", tags=["demands"])


# Pydantic models for request/response


class ClubDemandCreate(BaseModel):
    club_name: str
    club_contact_name: Optional[str] = None
    club_contact_email: Optional[str] = None
    deadline: Optional[str] = None  # ISO date string
    priority: str = "medium"
    position: list[str]
    age_min: Optional[int] = None
    age_max: Optional[int] = None
    nationality_preferences: Optional[list[str]] = None
    budget_min: Optional[int] = None
    budget_max: Optional[int] = None
    metrics_requirements: Optional[dict] = None
    playing_style_notes: Optional[str] = None
    default_season: str
    internal_notes: Optional[str] = None


class ClubDemandUpdate(BaseModel):
    club_name: Optional[str] = None
    club_contact_name: Optional[str] = None
    club_contact_email: Optional[str] = None
    deadline: Optional[str] = None
    status: Optional[str] = None
    priority: Optional[str] = None
    position: Optional[list[str]] = None
    age_min: Optional[int] = None
    age_max: Optional[int] = None
    nationality_preferences: Optional[list[str]] = None
    budget_min: Optional[int] = None
    budget_max: Optional[int] = None
    metrics_requirements: Optional[dict] = None
    playing_style_notes: Optional[str] = None
    default_season: Optional[str] = None
    internal_notes: Optional[str] = None


class MatchCounts(BaseModel):
    suggested: int = 0
    proposed: int = 0
    interested: int = 0
    rejected: int = 0
    signed: int = 0


class DemandSummary(BaseModel):
    id: int
    club_name: str
    position: list[str]
    status: str
    priority: str
    deadline: Optional[str]
    created_date: str
    match_counts: MatchCounts


class DemandDetail(BaseModel):
    id: int
    club_name: str
    club_contact_name: Optional[str]
    club_contact_email: Optional[str]
    created_date: str
    deadline: Optional[str]
    status: str
    priority: str
    position: list[str]
    age_min: Optional[int]
    age_max: Optional[int]
    nationality_preferences: Optional[list[str]]
    budget_min: Optional[int]
    budget_max: Optional[int]
    metrics_requirements: Optional[dict]
    playing_style_notes: Optional[str]
    default_season: str
    internal_notes: Optional[str]
    match_counts: MatchCounts


class PlayerInfo(BaseModel):
    id: int
    name: str
    team: Optional[str]
    league: Optional[str]
    season: Optional[str]
    position: Optional[str]
    age: Optional[int]
    nationality: Optional[str]
    image_url: Optional[str]
    market_value: Optional[float]
    key_metrics: dict[str, float]


class DemandMatchResponse(BaseModel):
    id: int
    demand_id: int
    player_season_id: int
    match_score: float
    status: str
    added_date: str
    status_updated_date: str
    notes: Optional[str]
    player: PlayerInfo


class DemandMatchUpdate(BaseModel):
    status: Optional[str] = None
    notes: Optional[str] = None


class MatchRequest(BaseModel):
    season: Optional[str] = None
    limit: int = 50


# Helper functions


def get_match_counts(db: Session, demand_id: int) -> MatchCounts:
    """Get counts of matches by status for a demand."""
    counts = (
        db.query(DemandMatch.status, func.count(DemandMatch.id))
        .filter(DemandMatch.demand_id == demand_id)
        .group_by(DemandMatch.status)
        .all()
    )

    match_counts = MatchCounts()
    for status, count in counts:
        setattr(match_counts, status, count)

    return match_counts


def demand_to_summary(demand: ClubDemand, db: Session) -> DemandSummary:
    """Convert ClubDemand to DemandSummary."""
    return DemandSummary(
        id=demand.id,
        club_name=demand.club_name,
        position=demand.position,
        status=demand.status,
        priority=demand.priority,
        deadline=demand.deadline.isoformat() if demand.deadline else None,
        created_date=demand.created_date.isoformat(),
        match_counts=get_match_counts(db, demand.id),
    )


def demand_to_detail(demand: ClubDemand, db: Session) -> DemandDetail:
    """Convert ClubDemand to DemandDetail."""
    return DemandDetail(
        id=demand.id,
        club_name=demand.club_name,
        club_contact_name=demand.club_contact_name,
        club_contact_email=demand.club_contact_email,
        created_date=demand.created_date.isoformat(),
        deadline=demand.deadline.isoformat() if demand.deadline else None,
        status=demand.status,
        priority=demand.priority,
        position=demand.position,
        age_min=demand.age_min,
        age_max=demand.age_max,
        nationality_preferences=demand.nationality_preferences,
        budget_min=demand.budget_min,
        budget_max=demand.budget_max,
        metrics_requirements=demand.metrics_requirements,
        playing_style_notes=demand.playing_style_notes,
        default_season=demand.default_season,
        internal_notes=demand.internal_notes,
        match_counts=get_match_counts(db, demand.id),
    )


def calculate_player_age(player_season: PlayerSeason) -> Optional[int]:
    """Calculate player's current age."""
    current_year = datetime.now().year
    if player_season.birth_date:
        age = current_year - player_season.birth_date.year
        if datetime.now().month < player_season.birth_date.month or (
            datetime.now().month == player_season.birth_date.month
            and datetime.now().day < player_season.birth_date.day
        ):
            age -= 1
        return age
    elif player_season.born_year:
        return current_year - player_season.born_year
    return None


def match_to_response(
    match: DemandMatch, db: Session, key_metrics: Optional[dict] = None
) -> DemandMatchResponse:
    """Convert DemandMatch to DemandMatchResponse with enriched player data."""
    player_season = match.player_season

    # Get team, league, season info
    team = db.query(Team).filter(Team.id == player_season.team_id).first()
    league = db.query(League).filter(League.id == player_season.league_id).first()
    season = db.query(Season).filter(Season.id == player_season.season_id).first()

    # Get key metrics if specified
    metrics_dict = {}
    if key_metrics:
        from scouting.db.models import MetricDefinition, PlayerMetric

        for metric_code in key_metrics.keys():
            metric_def = (
                db.query(MetricDefinition)
                .filter(MetricDefinition.code == metric_code)
                .first()
            )
            if metric_def:
                player_metric = (
                    db.query(PlayerMetric)
                    .filter(
                        and_(
                            PlayerMetric.player_season_id == player_season.id,
                            PlayerMetric.metric_id == metric_def.id,
                        )
                    )
                    .first()
                )
                if player_metric:
                    metrics_dict[metric_code] = player_metric.value

    player_info = PlayerInfo(
        id=player_season.id,
        name=player_season.player_name,
        team=team.name if team else None,
        league=league.name if league else None,
        season=season.label if season else None,
        position=player_season.position,
        age=calculate_player_age(player_season),
        nationality=player_season.nationality,
        image_url=player_season.image_url,
        market_value=player_season.market_value_eur,
        key_metrics=metrics_dict,
    )

    return DemandMatchResponse(
        id=match.id,
        demand_id=match.demand_id,
        player_season_id=match.player_season_id,
        match_score=match.match_score,
        status=match.status,
        added_date=match.added_date.isoformat(),
        status_updated_date=match.status_updated_date.isoformat(),
        notes=match.notes,
        player=player_info,
    )


# API Endpoints


@router.post("", response_model=DemandDetail, status_code=201)
def create_demand(demand: ClubDemandCreate, db: Session = Depends(get_db)):
    """Create a new club demand."""
    now = datetime.now()

    # Parse deadline if provided
    deadline = None
    if demand.deadline:
        try:
            deadline = datetime.fromisoformat(demand.deadline).date()
        except ValueError:
            raise HTTPException(status_code=400, detail="Invalid deadline format")

    # Create demand
    db_demand = ClubDemand(
        club_name=demand.club_name,
        club_contact_name=demand.club_contact_name,
        club_contact_email=demand.club_contact_email,
        created_date=now,
        deadline=deadline,
        status="open",
        priority=demand.priority,
        position=demand.position,
        age_min=demand.age_min,
        age_max=demand.age_max,
        nationality_preferences=demand.nationality_preferences,
        budget_min=demand.budget_min,
        budget_max=demand.budget_max,
        metrics_requirements=demand.metrics_requirements,
        playing_style_notes=demand.playing_style_notes,
        default_season=demand.default_season,
        internal_notes=demand.internal_notes,
    )

    db.add(db_demand)
    db.commit()
    db.refresh(db_demand)

    return demand_to_detail(db_demand, db)


@router.get("", response_model=list[DemandSummary])
def list_demands(
    status: Optional[str] = Query(None),
    club_name: Optional[str] = Query(None),
    position: Optional[str] = Query(None),
    date_from: Optional[str] = Query(None),
    date_to: Optional[str] = Query(None),
    db: Session = Depends(get_db),
):
    """List all demands with optional filters."""
    query = db.query(ClubDemand)

    if status:
        query = query.filter(ClubDemand.status == status)

    if club_name:
        query = query.filter(ClubDemand.club_name.ilike(f"%{club_name}%"))

    if position:
        query = query.filter(ClubDemand.position.contains([position]))

    if date_from:
        try:
            from_date = datetime.fromisoformat(date_from).date()
            query = query.filter(ClubDemand.created_date >= from_date)
        except ValueError:
            raise HTTPException(status_code=400, detail="Invalid date_from format")

    if date_to:
        try:
            to_date = datetime.fromisoformat(date_to).date()
            query = query.filter(ClubDemand.created_date <= to_date)
        except ValueError:
            raise HTTPException(status_code=400, detail="Invalid date_to format")

    demands = query.order_by(ClubDemand.created_date.desc()).all()
    return [demand_to_summary(d, db) for d in demands]


@router.get("/{demand_id}", response_model=DemandDetail)
def get_demand(demand_id: int, db: Session = Depends(get_db)):
    """Get demand detail."""
    demand = db.query(ClubDemand).filter(ClubDemand.id == demand_id).first()
    if not demand:
        raise HTTPException(status_code=404, detail="Demand not found")

    return demand_to_detail(demand, db)


@router.patch("/{demand_id}", response_model=DemandDetail)
def update_demand(
    demand_id: int, updates: ClubDemandUpdate, db: Session = Depends(get_db)
):
    """Update demand."""
    demand = db.query(ClubDemand).filter(ClubDemand.id == demand_id).first()
    if not demand:
        raise HTTPException(status_code=404, detail="Demand not found")

    # Update fields
    update_data = updates.model_dump(exclude_unset=True)

    # Handle deadline conversion
    if "deadline" in update_data and update_data["deadline"]:
        try:
            update_data["deadline"] = datetime.fromisoformat(
                update_data["deadline"]
            ).date()
        except ValueError:
            raise HTTPException(status_code=400, detail="Invalid deadline format")

    for field, value in update_data.items():
        setattr(demand, field, value)

    db.commit()
    db.refresh(demand)

    return demand_to_detail(demand, db)


@router.delete("/{demand_id}", status_code=204)
def delete_demand(demand_id: int, db: Session = Depends(get_db)):
    """Delete demand and associated matches."""
    demand = db.query(ClubDemand).filter(ClubDemand.id == demand_id).first()
    if not demand:
        raise HTTPException(status_code=404, detail="Demand not found")

    db.delete(demand)
    db.commit()
    return None


@router.post("/{demand_id}/match", response_model=list[DemandMatchResponse])
def run_matching(
    demand_id: int,
    match_request: Optional[MatchRequest] = None,
    db: Session = Depends(get_db),
):
    """Run matching algorithm for a demand."""
    demand = (
        db.query(ClubDemand)
        .options(joinedload(ClubDemand.matches))
        .filter(ClubDemand.id == demand_id)
        .first()
    )
    if not demand:
        raise HTTPException(status_code=404, detail="Demand not found")

    # Get matching parameters
    season = None
    limit = 50
    if match_request:
        season = match_request.season
        limit = match_request.limit

    # Find matching players
    matched_players = find_matching_players(db, demand, season=season, limit=limit)

    # Create/update demand matches
    matches = create_demand_matches(db, demand_id, matched_players)

    # Return enriched match data
    return [
        match_to_response(m, db, key_metrics=demand.metrics_requirements)
        for m in matches
    ]


@router.get("/{demand_id}/matches", response_model=list[DemandMatchResponse])
def get_demand_matches(
    demand_id: int,
    status: Optional[str] = Query(None),
    sort_by: str = Query("match_score", regex="^(match_score|added_date)$"),
    db: Session = Depends(get_db),
):
    """Get all matched players for a demand."""
    demand = db.query(ClubDemand).filter(ClubDemand.id == demand_id).first()
    if not demand:
        raise HTTPException(status_code=404, detail="Demand not found")

    query = (
        db.query(DemandMatch)
        .options(joinedload(DemandMatch.player_season))
        .filter(DemandMatch.demand_id == demand_id)
    )

    if status:
        query = query.filter(DemandMatch.status == status)

    # Sort
    if sort_by == "match_score":
        query = query.order_by(DemandMatch.match_score.desc())
    else:  # added_date
        query = query.order_by(DemandMatch.added_date.desc())

    matches = query.all()
    return [
        match_to_response(m, db, key_metrics=demand.metrics_requirements)
        for m in matches
    ]


@router.patch(
    "/{demand_id}/matches/{player_season_id}", response_model=DemandMatchResponse
)
def update_match(
    demand_id: int,
    player_season_id: int,
    updates: DemandMatchUpdate,
    db: Session = Depends(get_db),
):
    """Update match status and/or notes."""
    match = (
        db.query(DemandMatch)
        .filter(
            and_(
                DemandMatch.demand_id == demand_id,
                DemandMatch.player_season_id == player_season_id,
            )
        )
        .first()
    )
    if not match:
        raise HTTPException(status_code=404, detail="Match not found")

    # Update fields
    if updates.status:
        match.status = updates.status
        match.status_updated_date = datetime.now()
    if updates.notes is not None:
        match.notes = updates.notes

    db.commit()
    db.refresh(match)

    demand = db.query(ClubDemand).filter(ClubDemand.id == demand_id).first()
    return match_to_response(match, db, key_metrics=demand.metrics_requirements)


@router.post(
    "/{demand_id}/matches/{player_season_id}/propose",
    response_model=DemandMatchResponse,
)
def propose_match(
    demand_id: int, player_season_id: int, db: Session = Depends(get_db)
):
    """Shortcut to set status to proposed."""
    match = (
        db.query(DemandMatch)
        .filter(
            and_(
                DemandMatch.demand_id == demand_id,
                DemandMatch.player_season_id == player_season_id,
            )
        )
        .first()
    )
    if not match:
        raise HTTPException(status_code=404, detail="Match not found")

    match.status = "proposed"
    match.status_updated_date = datetime.now()

    db.commit()
    db.refresh(match)

    demand = db.query(ClubDemand).filter(ClubDemand.id == demand_id).first()
    return match_to_response(match, db, key_metrics=demand.metrics_requirements)


@router.delete("/{demand_id}/matches/{player_season_id}", status_code=204)
def delete_match(
    demand_id: int, player_season_id: int, db: Session = Depends(get_db)
):
    """Remove player from demand matches."""
    match = (
        db.query(DemandMatch)
        .filter(
            and_(
                DemandMatch.demand_id == demand_id,
                DemandMatch.player_season_id == player_season_id,
            )
        )
        .first()
    )
    if not match:
        raise HTTPException(status_code=404, detail="Match not found")

    db.delete(match)
    db.commit()
    return None
